include/seqan3/range/views/take.hpp

```diff
@@ -455,7 +443,7 @@ class view_take : public std::ranges::view_interface<view_take<urng_t, exactly,
      *
      * No-throw guarantee.
      */
-    constexpr size_type size() const noexcept
+    constexpr auto size() const noexcept
```

Comment:
Okay you are right:

* `target_size` is always declared as `size_t`.
* before it did implicit conversion to `seqan3::size_type_t<urng_t>`

I'll change that

Reply:
boah makes it ultra complicated. The problem is that a declaration like:

```c++
    constexpr std::ranges::range_size_t<urng_t> size() const noexcept
        requires exactly || std::ranges::sized_range<urng_t>
    {
        return target_size;
    }
```

will always fail for non sized ranges at instantiation of this class (all type are known, thus nothing will be deferred). 

Adding a template parameter to force the function into two-phase lookup

```c++
    template <typename urng_type = urng_t>
    constexpr std::ranges::range_size_t<urng_type> size() const noexcept
        requires exactly || std::ranges::sized_range<urng_type>
    {
        return target_size;
    }
```

does not work if `urng_type` is not sized. Because I `range_size_t` is only defined on sized ranges.

Splitting the function into two would work:

```c++
    template <typename urng_type = urng_t>
    constexpr size_t size() const noexcept
        requires exactly
    {
        return target_size;
    }

    template <typename urng_type = urng_t>
    constexpr std::ranges::range_size_t<urng_type> size() const noexcept
        requires !(exactly) && std::ranges::sized_range<urng_type>
    {
        return target_size;
    }
```

What to do?

Reply:
The standard does it like this:

```c++
    constexpr auto size() requires sized_range<V> {
      auto n = ranges::size(base_);
      return ranges::min(n, static_cast<decltype(n)>(count_));
    }
```

They cheat, because the function body will only be instantiated if the requires clause is `true`.

Thus auto would solve our issues here too, but it is bad for documentation.

Reply:
On the otherhand

```c++
constexpr auto begin() noexcept
```

does not say anything about the returned type.

Reply:
eww.. Ok this turned out nastier than I thought.

> The standard does it like this:
> ...

That's right but they also document the return type more with "Effects: Equivalent to: " (at least in the iota view I look at for an example), such that the user knows what to expect.

> On the otherhand
> `constexpr auto begin() noexcept`
> does not say anything about the returned type.

Just another flaw? :D (at least if there's no additional implementation)

Well, after looking at you solutions I'd propose to go with `auto` as the return type and document it? (e.g. `If urng_type is sized, return type is std::ranges::range_size_t<urng_type>, else size_t`)

Reply:
You forgot the `exactly` part, how about:

```c++
    /*!\brief Returns the number of elements in the view.
     * \returns The number of elements in the view.
     *
     * This overload is only available if the underlying range models std::ranges::sized_range (return type is std::ranges::range_size_t<urng_type>) or for
     * specialisation that have the `exactly` (return type is std::size_t) template parameter set.
     * If both conditions are true the return type is `std::size_t`.
     *
     * ### Complexity
     *
     * Constant.
     *
     * ### Exceptions
     *
     * No-throw guarantee.
     */
```